### PR TITLE
Dismissable should cull and clip background

### LINKED ISSUE #6127: Background for dismissable widget overflows on Nexus 5x
And this problem does not happen on 5, 6p or galaxy nexus screen.

Details are attached at: b/31798745

### PR DESCRIPTION
**Pull request by abarth**
When not dismissing, the Dismissable widget should cull its background.
When a dismiss is in progress, it should clip the background to just the
part that is revealed.

Fixes #6127

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie @HansMuller

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 1]
```diff
@@ -857,6 +857,13 @@ class RenderBackdropFilter extends RenderProxyBox {
 
 /// A class that provides custom clips.
 abstract class CustomClipper<T> {
+  /// Creates a custom clipper.
+  ///
+  /// The clipper will update its clip whenever [reclip] notifies its listeners.
+  const CustomClipper({ Listenable reclip }) : _reclip = reclip;
+
+  final Listenable _reclip;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: I think you should discuss reclip in the class description dartdocs (also maybe describe the general design of the class there), and in more detail in the constructor description.

File: packages/flutter/lib/src/widgets/dismissable.dart [hunk 2]
```diff
@@ -340,17 +378,29 @@ class _DismissableState extends State<Dismissable> with TickerProviderStateMixin
...
     if (background != null) {
-      backgroundAndChild = new Stack(
-        children: <Widget>[
-          new Positioned(left: 0.0, top: 0.0, bottom: 0.0, right: 0.0, child: background),
-          new Viewport(child: backgroundAndChild)
-        ]
-      );
+      List<Widget> children = <Widget>[];
+
+      if (!_moveAnimation.isDismissed) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/dismissable.dart]: won't this make the background suddenly disappear once the swipe away transition is complete?

**Comment by Hixie** [comment]
Please mail flutter-users about the renaming of shouldRepaint to shouldReclip.

test?

LGTM except for the build function of the dismissable, which I don't understand. RSLGTM on that though if it works.

File: packages/flutter/lib/src/widgets/dismissable.dart [hunk 3]
```diff
@@ -114,6 +114,43 @@ class Dismissable extends StatefulWidget {
...
+        return new Rect.fromLTRB(0.0, 0.0, size.width, offset);
+    }
+    return null;
+  }
+
+  Rect getApproximateClipRect(Size size) => getClip(size);
+
+  bool shouldReclip(_DismissableClipper oldClipper) {
+    return oldClipper.axis != axis
+        || oldClipper.moveAnimation != moveAnimation;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/dismissable.dart]: Good idea.  Doesn't actually matter in this specific case, but that's a better practice that we should model.

File: packages/flutter/lib/src/widgets/dismissable.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/dismissable.dart]: Ah, I see it now. That makes sense.

Should we put a global key around the background when we insert it? Or should we rely on the caller to do so if they have state in there that they want to preserve across the various ways we build this?

If you don't want to put a global key in this build function to do that, we should document that unless they use a global key in their background, it could get its state reset, and we should suggest putting the same global key on background and secondaryBackground too if they are structurally similar and should have state kept across them.

Comment (abarth) [inline packages/flutter/lib/src/widgets/dismissable.dart]: Yeah, I think we're not super clear about when a widget is responsible for keeping sync between the widgets you hand in.  In this case, it makes sense that the background will be culled out when its not on screen, so it's sensible that it wouldn't always sync.  The loss of sync between the swipeaway and the collapse is probably less expected through...

Comment (abarth) [inline packages/flutter/lib/src/widgets/dismissable.dart]: I realized the `Key` isn't needed for the foreground because it stays at the end of the list, which will be synced properly.

I'm going to leave the background issue for another patch because I'm not changing it here.
